flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataset/DataSetAggregateRule.scala

```diff
@@ -44,17 +45,24 @@ class DataSetAggregateRule
     val inputType = agg.getInput.getRowType()
 
     // add grouping fields, position keys in the input, and input type
-    val aggregateFunction = AggregateFactory.createAggregateInstance(agg.getAggCallList,
-        inputType, grouping)
+    val aggregateResult = AggregateUtil.createOperatorFunctionsForAggregates(agg,
+        inputType, rel.getRowType, grouping)
+
+    val mapNode = new DataSetMap(rel.getCluster,
+      traitSet,
+      convInput,
+      aggregateResult.intermediateDataType,
+      agg.toString,
+      aggregateResult.mapFunc)
 
     new DataSetGroupReduce(
       rel.getCluster,
       traitSet,
-      convInput,
+      mapNode,
       rel.getRowType,
       agg.toString,
-      grouping,
-      aggregateFunction)
+      (0 until grouping.length).toArray,
+      aggregateResult.reduceGroupFunc)
```

Comment:
Can we also update the `DataSetGroupReduce` to accept a function that generates the `GroupReduceFunction` instead of the `GroupReduceFunction` itself?